<h1>Algoritmos de Optimización - Seminario</h1><br>
<b>Nombres y apellidos:</b> Ricardo Triana<br>
<b>URL:</b> <a>https://github.com/ricardo-triana/03MAIR-Algoritmos-de-Optimizacion-2019/tree/master/SEMINARIO</a><br>
<b>Problema: </b>
Combinar cifras y operaciones<br>
Descripción del problema:<br>
El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.<br>
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/) <br>
• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 2:<br>
4+2-6/3*1 = 4<br>
Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:<br>
- ¿Qué valor máximo y mínimo a priori se pueden obtener según las condiciones?<br>
- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?<br>

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.

Respuesta

Lo primero es calcular cuántas posibles combinaciones se pueden tener obteniendo 5 elementos de un total de 9.<br>
Esta cifra indica cuántos conjuntos de 5 elementos distintos se pueden formar y se obtiene con la combinatoria de 9 y 5.
Según la fórmula de la combinatoria, el número de conjuntos es 9!/(5!(9-5)!) = 126.<br>
Por otra parte, cada uno de estos conjuntos puede ordenarse de 5! = 120 formas diferentes.
Y al adicionar los símbolos +, -, * y /, un conjunto de cuatro elementos, este conjunto puede organizarse de 4! = 24 formas distintas.
Por lo tanto, hay 126 * 120 * 24 = 362880 posibles soluciones.
Los valores serán números reales (algunos de ellos enteros) en un rango limitado por la máxima y mínima cantidad posible que arroje la combinación de números y signos.<br>
En este problema la restricción es que el valor ingresado debe ser entero. Como se explicará más adelante, entre las 362.880 posibles soluciones, hay 90.000 soluciones que arrojan valores enteros.

<h3>¿Qué valor máximo y mínimo a priori se pueden obtener según las condiciones?</h3>

El máximo debe optimizar los números más altos con las operaciones que permiten avanzar y buscar descontar lo mínimo posible dado que se deben usar todos los operadores: 9x8+7-1/6.
El mínimo debe buscar el menor número negativo posible a partir de una diferencia que de negativa y se amplie mediante multiplicación. A este resultado, se le suma la menor fracción posible con los números restantes: 1-9x8+2/7.

In [1]:
print('Cálculo del valor máximo:')
print(eval('9*8+7-1/6'))

Cálculo del valor máximo:
78.83333333333333


In [2]:
print('Cálculo del valor mínimo:')
print(eval('1-9*8+2/7'))

Cálculo del valor mínimo:
-70.71428571428571


<h3>¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo?</h3>

No, un contraejemplo sería el valor 78, no puede ser hallado. O el valor -70. Los valores enteros cerca a los extremos. En el archivo adjunto, soluciones.xlsx se han guardado todas las posibles soluciones calculadas mediante el método de fuerza bruta. En este mismo archivo se ha hecho un conteo para determinar la cantidad de posibles valores que se pueden obtener (3533) y la cantidad de ocurrencias de cada uno de ellos. Hay números enteros que pueden representarse de varias maneras diferentes, sin embargo, algunos como los mencionados anteriormente no tienen solución. El 5 por ejemplo, puede lograrse de 2.880 formas diferentes, mientras que valores como el 78 o el -70 no se pueden lograr mediante combinación alguna.

Modelo para el espacio de soluciones
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguméntalo)

La estructura de datos que mejor se adapta al problema es una lista con elementos de tipo string unitario (cada uno con un solo operador o número). Preferí elegir esta estructura por la facilidad de usar índices para recorrerlos, por la disponibilidad de múltiples funciones como <b>"shuffle"</b>, <b>"append"</b>, entre otras. De esta forma, también se evitan posibles problemas que podría generar el intérprete de Python al resolver operaciones como 123+45, en el que se busca que la solución sea 12345 y no 168. 

Según el modelo para el espacio de soluciones
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Según el modelo, la función objetivo es eval(expresion) - esperado = 0, donde expresion es una posible solución y <b>esperado</b> es el valor que el usuario del programa ingresa para encontrar una expresión que lo genere.
No es un problema de maximización ni de minimización. De hecho, a medida que se avanza en la solución, esta puede alejarse o acercarse al valor de entrada según el orden de los operadores o los números. No se puede hablar de convergencia de las soluciones. Es un problema de búsqueda en realidad.

Diseña un algoritmo para resolver el problema por fuerza bruta

La resolución del algoritmo mediante fuerza bruta es dispendiosa porque requiere calcular todas las posibles 362.880 soluciones. Hay que calcular todos los posibles subconjuntos de 5 elementos a partir de la lista de 9 y para cada uno de ellos, calcular todas las posibles permutaciones (cambiando el orden dentro de cada subconjunto). Cada una de ellas, se debe poder combinar a su vez, con todas las posibles permutaciones del conjunto de operadores.<br>
Por lo tanto, la solución al problema se divide en varios algoritmos necesarios para la función generadora de soluciones. La primera función, <b>subconjuntos</b> permite obtener todos los posibles subconjuntos de 5 elementos a partir de una lista con los números de 1 al 9.<br>
Ya que el problema exige probar con los números en cualquier orden, se tiene la función <b>permutaciones</b> para calcular todas las posibles variantes (en cuanto a orden) de un conjunto determinado. Esta función se aplica tanto para los subconjuntos de números como para el conjunto de operadores.
Finalmente, hay una rutina principal que hace las combinaciones de cada permutación numérica con cada permutación de operadores en una única lista para aplicar la función de evaluación.

In [3]:
#Algoritmo para obtener todos los posibles subconjuntos de 5 elementos a partir del 
#conjunto lista = ['1','2','3','4','5','6','7','8','9']

#Los resultados se van agregando a una lista de listas 
lista_subconjuntos = []

def subconjuntos(L, n, k, indice, temp, i): 
    if(indice == k): 
        #Se almacena cada nuevo número en una lista parcial
        parcial = []
        for j in range(k): 
            parcial.append(temp[j])
        #Cuando la lista parcial se ha recorrido completamente, se agrega a la lista de listas
        lista_subconjuntos.append(parcial)
        return
    else:
        #Mientras el contador i sea menor que la longitud de la lista de elementos original
        if(i < n): 
            #Calcula el siguiente elemento y lo agrega a la lista temporal
            temp[indice] = L[i]
            #Llama de manera recursiva la función para indice + 1 e indice
            subconjuntos(L, n, k, indice + 1, temp, i + 1) 
            subconjuntos(L, n, k, indice, temp, i + 1)
            return #resultado

lista = [str(i+1) for i in range(9)] 
  
k = 5
n = len(lista) 
datos = list(range(k))
subconjuntos(lista, n, k, 0, datos, 0)
print(lista_subconjuntos)

[['1', '2', '3', '4', '5'], ['1', '2', '3', '4', '6'], ['1', '2', '3', '4', '7'], ['1', '2', '3', '4', '8'], ['1', '2', '3', '4', '9'], ['1', '2', '3', '5', '6'], ['1', '2', '3', '5', '7'], ['1', '2', '3', '5', '8'], ['1', '2', '3', '5', '9'], ['1', '2', '3', '6', '7'], ['1', '2', '3', '6', '8'], ['1', '2', '3', '6', '9'], ['1', '2', '3', '7', '8'], ['1', '2', '3', '7', '9'], ['1', '2', '3', '8', '9'], ['1', '2', '4', '5', '6'], ['1', '2', '4', '5', '7'], ['1', '2', '4', '5', '8'], ['1', '2', '4', '5', '9'], ['1', '2', '4', '6', '7'], ['1', '2', '4', '6', '8'], ['1', '2', '4', '6', '9'], ['1', '2', '4', '7', '8'], ['1', '2', '4', '7', '9'], ['1', '2', '4', '8', '9'], ['1', '2', '5', '6', '7'], ['1', '2', '5', '6', '8'], ['1', '2', '5', '6', '9'], ['1', '2', '5', '7', '8'], ['1', '2', '5', '7', '9'], ['1', '2', '5', '8', '9'], ['1', '2', '6', '7', '8'], ['1', '2', '6', '7', '9'], ['1', '2', '6', '8', '9'], ['1', '2', '7', '8', '9'], ['1', '3', '4', '5', '6'], ['1', '3', '4', '5', '7'], 

In [4]:
#Se calcula el número de posibles subconjuntos de 5 elementos a partir de la lista de 9 elementos, para verificar
#si el cálculo teórico es correcto n!/(k!(n-k)!) =  9!/(5!(9-5)!) = 126.
print('La cantidad de subconjuntos de ' + str(k) + ' elementos a partir de un conjunto de ' + str (n) + ' elementos es:')
print(len(lista_subconjuntos))

La cantidad de subconjuntos de 5 elementos a partir de un conjunto de 9 elementos es:
126


<b>Algoritmo para obtener todas las posibles permutaciones de una lista</b><br>
Para cada uno de los subconjuntos, se debe probar con todas las posibles permutaciones.<br>
El siguiente algoritmo recibe por entrada una lista y entrega una lista de listas con todas las permutaciones.<br>
Este se usara tanto para cada uno de los subconjuntos del paso anterior, así como con la lista de operadores matermáticos.

In [5]:
#Algoritmo para la obtención de todas las posibles permutaciones de una lista de n elementos
#Recibe una lista y devuelve una lista de listas
n = 5
lista = [str(i+1) for i in range(n)]

def permutaciones(L):
    if len(L)==2:
        resultado = [[L[0],L[1]],[L[1],L[0]]]
    else:
        resultado = []
        for i in range(len(L)):
            A = L[:i] + L[i+1:]
            perm = permutaciones(A) #Procedimiento recursivo en el que A es L menos un elemento diferente cada vez
            for j in range(len(perm)):
                res = [L[i]] + perm[j]
                resultado.append(res)
    return resultado


In [6]:
#Se verifica que la cantidad de permutaciones coincida con el valor calculado teóricamente.
#Para un conjunto de n elementos, debe ser igual a n!. Por lo tanto, para cada subconjunto de 5 elementos, 
#hay 5! = 120 posibilidades y para los operadores, hay 4! = 24 posibilidades.
conjunto_prueba = [str(i+1) for i in range(5)]
operadores = ['+','-','*','/']

perm_conj_prueba = permutaciones(conjunto_prueba)
perm_conj_oper = permutaciones(operadores)

print('La cantidad de permutaciones del conjunto de prueba es: ')
print(len(perm_conj_prueba))
print('La cantidad de permutaciones del conjunto de operadores es: ')
print(len(perm_conj_oper))


La cantidad de permutaciones del conjunto de prueba es: 
120
La cantidad de permutaciones del conjunto de operadores es: 
24


Para calcular todas las posibles soluciones, el programa principal debe recorrer la variable <b>lista_subconjuntos</b> generada por la función <b>subconjuntos</b> y para cada uno de los elementos de la lista, que es a su vez una lista de 5 elementos, aplicar la función <b>permutaciones</b> que devuelve todas las posibles permutaciones para cada lista. Mediante un bucle adicional, se deben combinar cada permutación de números con cada una de las permutaciones de operadores, que se encuentran en la variable <b>perm_conj_oper</b>.
Esto genera 126x120x24 = 362.880 soluciones diferentes.
Aunque bastaría con detener el programa cuando una solución sea hallada, se van a calcular todas las soluciones para efectos de analizar el universo de soluciones en detalle y para calcular con exactitud el mayor y el menor resultado posibles. Hacer esto, también permite conocer las diferentes soluciones para un mismo valor, si es que las tiene.

<h2>Programa principal para resolución por fuerza bruta<h2>

In [7]:
operadores = ['+','-','*','/']
perm_conj_oper = permutaciones(operadores)
#Se usa la variable lista_subconjuntos calculada en el primer algoritmo de este notebook, que contiene los 126 posibles
#subconjuntos de 5 elementos que se genera a partir de lista = ['1','2','3','4','5','6','7','8','9']

esperado = 2 #Este es el valor para el cual se van a buscar soluciones.
valor_minimo_posible = 0
valor_maximo_posible = 0

soluciones = []
valores = []
f = open("soluciones.csv","w+")

contador = 0

for x in lista_subconjuntos:
    permutaciones_x = permutaciones(x)
    for y in permutaciones_x:
        for z in perm_conj_oper:
            #y es una lista de 5 números
            #z es una lista de 4 operadores
            #Se crea una lista vacía de 9 elementos que se irá poblando con números y signos alternadamente
            expresion = ['' for j in range(9)]
            for i in range(5):
                pos = 2*i
                expresion[pos] = y[i]
                if (pos+1)<len(expresion):
                    expresion[pos+1] = z[i]
            cadena = ''.join(expresion)
            evaluacion = eval(cadena)
            #Se agregan todas las soluciones y los valores respectivos para su posterior análisis 
            soluciones.append(cadena)
            valores.append(evaluacion)
            linea = str(contador) + ';' + cadena + ";" + str(evaluacion) + '\n'
            #print(contador)
            f.write(linea)
            #Se cuentan las iteraciones
            contador += 1 
            if evaluacion > valor_maximo_posible:
                valor_maximo_posible = evaluacion
            if evaluacion < valor_minimo_posible:
                valor_minimo_posible = evaluacion
            
            if eval(cadena)==esperado:
                print('La solución es:' + cadena + ' = ' + str(esperado))

f.close
print('Valor máximo encontrado: ' + str(valor_maximo_posible))
print('Valor mínimo encontrado: ' + str(valor_minimo_posible))
print('Número de iteraciones: ' + str(contador))


La solución es:1+3/2*4-5 = 2
La solución es:1+3*4/2-5 = 2
La solución es:1+4/2*3-5 = 2
La solución es:1+4*3/2-5 = 2
La solución es:1-5+3/2*4 = 2
La solución es:1-5+3*4/2 = 2
La solución es:1-5+4/2*3 = 2
La solución es:1-5+4*3/2 = 2
La solución es:3/2*4+1-5 = 2
La solución es:3/2*4-5+1 = 2
La solución es:3*4/2+1-5 = 2
La solución es:3*4/2-5+1 = 2
La solución es:4/2*3+1-5 = 2
La solución es:4/2*3-5+1 = 2
La solución es:4*3/2+1-5 = 2
La solución es:4*3/2-5+1 = 2
La solución es:1/2*6+3-4 = 2
La solución es:1/2*6-4+3 = 2
La solución es:1*3-4+6/2 = 2
La solución es:1*3+6/2-4 = 2
La solución es:1*6/2+3-4 = 2
La solución es:1*6/2-4+3 = 2
La solución es:3+1/2*6-4 = 2
La solución es:3-1*4+6/2 = 2
La solución es:3*1-4+6/2 = 2
La solución es:3+1*6/2-4 = 2
La solución es:3*1+6/2-4 = 2
La solución es:3-4+1/2*6 = 2
La solución es:3-4+1*6/2 = 2
La solución es:3-4*1+6/2 = 2
La solución es:3-4+6*1/2 = 2
La solución es:3-4+6/2*1 = 2
La solución es:3+6*1/2-4 = 2
La solución es:3+6/2-1*4 = 2
La solución es

La solución es:1-2/3*9+7 = 2
La solución es:1-2*9/3+7 = 2
La solución es:1+7-2/3*9 = 2
La solución es:1+7-2*9/3 = 2
La solución es:1+7-9*2/3 = 2
La solución es:1+7-9/3*2 = 2
La solución es:1-9*2/3+7 = 2
La solución es:1-9/3*2+7 = 2
La solución es:7+1-2/3*9 = 2
La solución es:7+1-2*9/3 = 2
La solución es:7+1-9*2/3 = 2
La solución es:7+1-9/3*2 = 2
La solución es:7-2/3*9+1 = 2
La solución es:7-2*9/3+1 = 2
La solución es:7-9*2/3+1 = 2
La solución es:7-9/3*2+1 = 2
La solución es:1/2*6+4-5 = 2
La solución es:1/2*6-5+4 = 2
La solución es:1*4-5+6/2 = 2
La solución es:1*4+6/2-5 = 2
La solución es:1*6/2+4-5 = 2
La solución es:1*6/2-5+4 = 2
La solución es:4+1/2*6-5 = 2
La solución es:4-1*5+6/2 = 2
La solución es:4*1-5+6/2 = 2
La solución es:4+1*6/2-5 = 2
La solución es:4*1+6/2-5 = 2
La solución es:4-5+1/2*6 = 2
La solución es:4-5+1*6/2 = 2
La solución es:4-5*1+6/2 = 2
La solución es:4-5+6*1/2 = 2
La solución es:4-5+6/2*1 = 2
La solución es:4+6*1/2-5 = 2
La solución es:4+6/2-1*5 = 2
La solución es

La solución es:1+3/4*8-5 = 2
La solución es:1+3*8/4-5 = 2
La solución es:1-5+3/4*8 = 2
La solución es:1-5+3*8/4 = 2
La solución es:1-5+8*3/4 = 2
La solución es:1-5+8/4*3 = 2
La solución es:1+8*3/4-5 = 2
La solución es:1+8/4*3-5 = 2
La solución es:3/4*8+1-5 = 2
La solución es:3/4*8-5+1 = 2
La solución es:3*8/4+1-5 = 2
La solución es:3*8/4-5+1 = 2
La solución es:8*3/4+1-5 = 2
La solución es:8*3/4-5+1 = 2
La solución es:8/4*3+1-5 = 2
La solución es:8/4*3-5+1 = 2
La solución es:1/3*9+4-5 = 2
La solución es:1/3*9-5+4 = 2
La solución es:1*4-5+9/3 = 2
La solución es:1*4+9/3-5 = 2
La solución es:1*9/3+4-5 = 2
La solución es:1*9/3-5+4 = 2
La solución es:4+1/3*9-5 = 2
La solución es:4-1*5+9/3 = 2
La solución es:4*1-5+9/3 = 2
La solución es:4+1*9/3-5 = 2
La solución es:4*1+9/3-5 = 2
La solución es:4-5+1/3*9 = 2
La solución es:4-5+1*9/3 = 2
La solución es:4-5*1+9/3 = 2
La solución es:4-5+9*1/3 = 2
La solución es:4-5+9/3*1 = 2
La solución es:4+9*1/3-5 = 2
La solución es:4+9/3-1*5 = 2
La solución es

La solución es:7+1/3*9-8 = 2
La solución es:7-1*8+9/3 = 2
La solución es:7*1-8+9/3 = 2
La solución es:7+1*9/3-8 = 2
La solución es:7*1+9/3-8 = 2
La solución es:7-8+1/3*9 = 2
La solución es:7-8+1*9/3 = 2
La solución es:7-8*1+9/3 = 2
La solución es:7-8+9*1/3 = 2
La solución es:7-8+9/3*1 = 2
La solución es:7+9*1/3-8 = 2
La solución es:7+9/3-1*8 = 2
La solución es:7+9/3*1-8 = 2
La solución es:7+9/3-8*1 = 2
La solución es:9*1/3+7-8 = 2
La solución es:9*1/3-8+7 = 2
La solución es:9/3+1*7-8 = 2
La solución es:9/3*1+7-8 = 2
La solución es:9/3-1*8+7 = 2
La solución es:9/3*1-8+7 = 2
La solución es:9/3+7-1*8 = 2
La solución es:9/3+7*1-8 = 2
La solución es:9/3+7-8*1 = 2
La solución es:9/3-8+1*7 = 2
La solución es:9/3-8*1+7 = 2
La solución es:9/3-8+7*1 = 2
La solución es:1+4/6*9-5 = 2
La solución es:1+4*9/6-5 = 2
La solución es:1-5+4/6*9 = 2
La solución es:1-5+4*9/6 = 2
La solución es:1-5+9*4/6 = 2
La solución es:1-5+9/6*4 = 2
La solución es:1+9*4/6-5 = 2
La solución es:1+9/6*4-5 = 2
La solución es

La solución es:7-2*4+9/3 = 2
La solución es:7-4*2+9/3 = 2
La solución es:7+9/3-2*4 = 2
La solución es:7+9/3-4*2 = 2
La solución es:9/3-2*4+7 = 2
La solución es:9/3-4*2+7 = 2
La solución es:9/3+7-2*4 = 2
La solución es:9/3+7-4*2 = 2
La solución es:2/3*9+4-8 = 2
La solución es:2/3*9-8+4 = 2
La solución es:2*9/3+4-8 = 2
La solución es:2*9/3-8+4 = 2
La solución es:4+2/3*9-8 = 2
La solución es:4+2*9/3-8 = 2
La solución es:4-8+2/3*9 = 2
La solución es:4-8+2*9/3 = 2
La solución es:4-8+9*2/3 = 2
La solución es:4-8+9/3*2 = 2
La solución es:4+9*2/3-8 = 2
La solución es:4+9/3*2-8 = 2
La solución es:9*2/3+4-8 = 2
La solución es:9*2/3-8+4 = 2
La solución es:9/3*2+4-8 = 2
La solución es:9/3*2-8+4 = 2
La solución es:2/3*6+5-7 = 2
La solución es:2/3*6-7+5 = 2
La solución es:2*6/3+5-7 = 2
La solución es:2*6/3-7+5 = 2
La solución es:5+2/3*6-7 = 2
La solución es:5+2*6/3-7 = 2
La solución es:5+6*2/3-7 = 2
La solución es:5+6/3*2-7 = 2
La solución es:5-7+2/3*6 = 2
La solución es:5-7+2*6/3 = 2
La solución es

La solución es:3/4*8+5-9 = 2
La solución es:3/4*8-9+5 = 2
La solución es:3-5/4*8+9 = 2
La solución es:3-5*8/4+9 = 2
La solución es:3-8/4*5+9 = 2
La solución es:3*8/4+5-9 = 2
La solución es:3*8/4-9+5 = 2
La solución es:3-8*5/4+9 = 2
La solución es:3+9-5/4*8 = 2
La solución es:3+9-5*8/4 = 2
La solución es:3+9-8/4*5 = 2
La solución es:3+9-8*5/4 = 2
La solución es:5+3/4*8-9 = 2
La solución es:5+3*8/4-9 = 2
La solución es:5+8*3/4-9 = 2
La solución es:5+8/4*3-9 = 2
La solución es:5-9+3/4*8 = 2
La solución es:5-9+3*8/4 = 2
La solución es:5-9+8*3/4 = 2
La solución es:5-9+8/4*3 = 2
La solución es:8*3/4+5-9 = 2
La solución es:8*3/4-9+5 = 2
La solución es:8/4*3+5-9 = 2
La solución es:8/4*3-9+5 = 2
La solución es:9+3-5/4*8 = 2
La solución es:9+3-5*8/4 = 2
La solución es:9+3-8/4*5 = 2
La solución es:9+3-8*5/4 = 2
La solución es:9-5/4*8+3 = 2
La solución es:9-5*8/4+3 = 2
La solución es:9-8/4*5+3 = 2
La solución es:9-8*5/4+3 = 2
La solución es:3+4/6*9-7 = 2
La solución es:3+4*9/6-7 = 2
La solución es

Calcula la complejidad del algoritmo por fuerza bruta

El algoritmo tiene una complejidad definida por una función O(n!) debido a que tanto el cálculo de los posibles subconjuntos como las posibles permutaciones dependen del factorial del número de elementos.
Haciendo el análisis de OE de la función <b>permutaciones</b>, se llega a la conclusión que es una función de la forma T(n) = n^3 + 3n^2 + n (1 + T(n-1)) + 3. Como se puede ver, queda un término de la forma nT(n-1) y la función que es recurrente, se llama a si misma al menos 4 veces. Con cada llamada este n se va multiplicando por n-1, n-2, es decir, n!.

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

<h1>Resolución del problema mediante un algoritmo heurístico</h1><br>
A diferencia del método de fuerza bruta, es posible hallar una solución para varios casos en una fracción del tiempo

In [53]:
import random

def hallar_solucion(esperado, limite):
    #El primer argumento es el valor para el que se debe hallar solución
    #El segundo argumento es el número máximo de iteraciones para buscar solución
    conj = ['1','2','3','4','5','6','7','8','9']
    oper = ['+','-','*','/']

    max = 0
    min = 0

    if (esperado > 78.83333333333333) or (esperado < -70.71428571428571):
        print('El valor se encuentra por fuera del rango posible [-70.71428571428571, 78.33333333333333]')
    elif not(esperado==int(esperado)):
        print('El valor debe ser entero')
    else: 
        resultado = esperado + 1
        contador = 0
        while (not(resultado == esperado)) and (contador<limite):
            random.shuffle(conj)
            random.shuffle(oper)
            expresion = ['' for x in range(9)]
            for i in range(5):
                pos = 2*i
                expresion[pos] = conj[i]
                if (pos+1)<len(expresion):
                    expresion[pos+1] = oper[i]
            cadena = ''.join(expresion)
            resultado = eval(cadena)
            if resultado > max:
                max = resultado
            if resultado < min:
                min = resultado
            contador += 1
        if eval(cadena)==esperado:
            print('La solución es:' + cadena + ' = ' + str(esperado))
        else:
            print('No se encontró solución después de ' + str(limite) + ' iteraciones.')
        print('Valor máximo encontrado: ' + str(max))
        print('Valor mínimo encontrado: ' + str(min))
        print('Número de iteraciones: ' + str(contador))
        
valor = 5
hallar_solucion(valor, 10000)

La solución es:7/1+2*3-8 = 5
Valor máximo encontrado: 73.14285714285714
Valor mínimo encontrado: -55.5
Número de iteraciones: 214


Como se puede observar en el ejemplo anterior, en algunos casos el algoritmo es capaz de encontrar soluciones muy rápidamente, aunque esto obviamente es completamente aleatorio. En el peor de los casos, es posible que el algoritmo no encuentre una solución dentro del número de iteraciones, pero este número no depende directamente del tamaño de los conjuntos de números u operadores. Depende del valor máximo de iteraciones que se haya definido. Considerando que las 362.880 posibles soluciones se concentran en 3533 valores distintos, y de las 362.880 posibles soluciones solo 90.000 arrojan valores enteros que se pueden mapear en 147 números enteros distintos, calculamos el número mínimo de iteraciones para tener una probabilidad específica de encontrar una solución a partir de un valor dado, así:

In [54]:
from math import log
print('Probabilidad\tIteraciones i>\n')
for i in range(5):
    p = 0.5 + i/10
    print(str(p) + '\t\t' + str(log(1-p)/log(1-1/362280)))

Probabilidad	Iteraciones i>

0.5		251113.01399808633
0.6		331953.3481959138
0.7		436174.6655620528
0.8		583066.3621940002
0.9		834179.3761920866


Esta solución supone el peor caso, en el que un valor sólo tenga una solución asociada. En el análisis que se hace en el archivo soluciones.xlsx se encontró que entre los 3533 posibles valores que toma la evaluación de la función, sólo 147 son números enteros, pero por ejemplo, el 5 se puede representar de 2880 maneras diferentes, mientras que el -69 sólo tiene 16 posibles soluciones (y es entre los enteros, el que menos soluciones tiene). Esto indica que la probabilidad de cada valor es distinta y el peor caso sería el -69, cuya probabilidad es de 16/362880 

In [55]:
print('Probabilidad solución entera\tIteraciones i>\n')
for i in range(5):
    p = 0.5 + i/10
    print(str(p) + '\t\t' + str(log(1-p)/log(1-16/362880)))

Probabilidad solución entera	Iteraciones i>

0.5		15720.23147896684
0.6		20781.015650178975
0.7		27305.501211293933
0.8		36501.24712914582
0.9		52221.478608112666


Dado un valor entero, en el peor escenario se necesitarían 52222 iteraciones para tener una probabilidad de hallarlo del 90%.

(*)Calcula la complejidad del algoritmo

Dado que el algoritmo incorpora la función shuffle de Python, es necesario ver su implementación, la cual se detalla a continuación:

In [28]:
def shuffle(self, x, random=None, int=int):
    """x, random=random.random -> shuffle list x in place; return None.
    Optional arg random is a 0-argument function returning a random
    float in [0.0, 1.0); by default, the standard random.random.
    """
    if random is None:
        random = self.random
        for i in reversed(xrange(1, len(x))):
            # pick an element in x[:i+1] with which to exchange x[i]
            j = int(random() * (i+1))
            x[i], x[j] = x[j], x[i]

Esta función es de orden <b>n</b>. Llama internamente a la función <b>reversed</b>, que simplemente invierte el orden de iteración del rango dado. Por lo tanto, La máxima complejidad del algoritmo depende la cantidad de iteraciones, multiplicada por n más un número de operaciones elementales.
La cantidad de operaciones es una función del tipo i*(an + c), luego es de O(n), donde i es el número de iteraciones, a es la cantidad de veces que se corre un bucle simple y c un número de operaciones elementales. También se usa la función <b>len</b>, la cual se estima que es de O(n) ya que debe recorrer la lista una vez para obtener la cantidad de elementos de esta.   

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

En este caso, se necesita generar un número entero y aleatorio en el rango comprendido entre -70.7142857142857 y 78.8333333. Como estos límites no son enteros, una función para generar un número aleatorio y entero en un rango determinado en Python se llama <b>radint</b>, de la librería <b>random</b>. A continuación se hace una prueba para generar un valor aleatorio de entrada:

In [59]:
from random import randint
valor = randint(-70, 78)
print(valor)
iteraciones = 52222

-23


In [60]:
#Prueba de la función con la entrada aleatoria
hallar_solucion(valor, iteraciones)

La solución es:3-4/1*8+6 = -23
Valor máximo encontrado: 78.33333333333333
Valor mínimo encontrado: -70.4
Número de iteraciones: 2420


In [ ]:
Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Referencias:

Combinatoria (definición), Wikipedia <a>https://es.wikipedia.org/wiki/Combinatoria</a><br>
Manual de la asignatura - Matemáticas para la Inteligencia Artificial, Dr. Gherardo Varando.<br>
Presentación VC5 – Algoritmos Heurísticos de la asignatura Algoritmos de Optimización, Dr. Raúl Reyero<br>
Implementación de la función random.shuffle: <a>http://svn.python.org/projects/python/trunk/Lib/random.py</a><br>
Valores aleatorios: la biblioteca random
<a>http://www.mclibre.org/consultar/python/lecciones/python-biblioteca-random.html</a>

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Hay una dificultad en el cálculo del número de iteraciones adecuado para hallar soluciones. En este caso, este número se calculó de forma teórica tras haber analizado en detalle el universo de soluciones. Pero si el problema fuera de una dimensión mayor (por ejemplo con más dígitos y más operadores), pienso que sería necesario hallar el número de iteraciones observando si el algoritmo en general es exitoso o no en encontrar soluciones, ya que el universo de soluciones mismo fue hallado mediante fuerza bruta y no mediante este algoritmo, pero en dimensiones mayores, podría no ser calculable o requeriría muchas horas de ejecución así como mucha memoria para ir almacenando los datos. Con tamaños mayores, también se ampliaría el rango de los posibles valores que toman las soluciones. Al menos, sería necesario calcular teóricamente esos nuevos límites para poder saber si al menos es posible obtener una solución para el valor de entrada. Por otra parte, el universo de 362.880 con sólo un elemento adicional ya sobrepasaría el millón de posibles soluciones, dado que las combinatorias y las permutaciones requeridas son funciones basadas en factoriales. La ventaja es que las funciones creadas en esta solución heurística, permiten fácilmente operar con conjuntos mayores. (De hecho también las funciones de cálculo de subconjuntos y permutaciones elaboradas para el cálculo por fuerza bruta pueden funcionar con tamaños mayores, asi fueron diseñadas).